lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,354 @@
+      end
+
+      def verify(credit_card, options = {})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r(("Number\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
+          gsub(%r(("expirationDate\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
+          gsub(%r(("FirstName\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
+          gsub(%r(("LastName\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]')
+      end
+
+      private
+
+      def add_clerk(post, options)
+        post[:clerk] = {}
+        post[:clerk][:numericId] = options[:numeric_id]
+      end
+
+      def add_invoice(post, money, options)
+        post[:amount] = {}
+        post[:amount][:total] = money.to_f
+        post[:amount][:tax] = options[:amount][:tax].to_f || 0.0
+      end
+
+      def api_options(post, options)
+        post[:apiOptions] = options[:api_options] if options[:api_options].present?
+      end
+
+      def add_credentials(post, options)
+        post[:dateTime] = options[:date_time] || current_date_time
+        post[:credential] = {}
+        post[:credential][:clientGuid] = options[:credential][:client_guid]
+        post[:credential][:authToken] = options[:credential][:auth_token]
+      end
+
+      def add_datetime(post, options)
+        post[:dateTime] = options[:date_time] || current_date_time
+      end
+
+      def add_transaction(post, options)
+        add_purchase_card(post, options)
+        add_override_bus_date(post, options)
+        add_amex(post, options)
+        add_auto(post, options)
+        add_card_on_file(post, options)
+        add_hotel(post, options)
+        add_pin(post, options)
+      end
+
+      def add_device(post, options)
+        if options[:device].present?
+          post[:device] = {}
+          post[:device][:manufacturer] = options[:device][:manufacturer]
+          post[:device][:model] = options[:device][:model]
+          post[:device][:promptPostalCode] = options[:device][:prompt_postal_code]
+          post[:device][:promptSecurityCode] = options[:device][:prompt_security_code]
+          post[:device][:promptStreetNumber] = options[:device][:prompt_street_number]
+          post[:device][:terminalId] = options[:device][:terminal_id]
+        end
+      end
+
+      def add_emv(post, options)
+        if options[:emv].present?
+          post[:emv] = {}
+          post[:emv][:tlvData] = options[:emv][:tlv_data]
+        end
+      end
+
+      def add_p2pe(post, options)
+        if options[:p2pe].present?
+          post[:p2pe] = {}
+          post[:p2pe][:format] = options[:p2pe][:format]
+          post[:p2pe][:ksn] = options[:p2pe][:ksn]
+        end
+      end
+
+      def add_card(post, credit_card, options)
```

Comment:
I think this method could use a slight refactor. The first conditional block is fine, though I would use the suggested check below (`is_a?(CreditCard)`) to add the necessary information to the post body, and then for additional fields, you can simply append them to the hash with simple one liners outside of the conditional block.

For lines `187-188` and `196-197`, you're adding the same option to `post` but expecting the values to be in different places in `options`. I think `187-188` is the way we should handle it, unless we find that we are handling a lot of extra `card` metadata that becomes cumbersome to manage on an individual field basis.

Similarly, ln `203-207` are doing what `198-200` are doing. As I mentioned in my other comment however, we should grab `security_value` from the `CreditCard` object as well as other `CreditCard` attributes (expiration `year`, `month`, etc.).

Since this is a lot, here's some pseudocode of what I'm thinking a refactor might look like:

```ruby
if CreditCard
  post[card_num] = CreditCard.number
  post[exp_mon] = CreditCard.month
  post[exp_year] = CreditCard.year
  post[present] = options[present]
else
  post[token] = token

# additional information not in the CreditCard object
return
```